### ISSUE TITLE
SQL query combination

### ISSUE THREAD

**Issue by Jannis** [graphprotocol/graph-node#857]

**Comment by lutter** [comment]
The rough plan of attack is following what's described in [this blog post](https://blog.hasura.io/architecture-of-a-high-performance-graphql-to-sql-server-58d9944b8a87/) I have to look into what that would look like in our case, and break it into more fine-grained tasks.

**Comment by lutter** [comment]
I just did a quick test to better gauge the speedup we can get from this. I took the following GraphQL query, which is slow in production:
```graphql
query TranscodersQuery($_v0_skip: Int, $_v1_first: Int) {
  transcoders(skip: $_v0_skip, first: $_v1_first) {
    ...TranscoderFragment
    __typename
  }
}

fragment TranscoderFragment on Transcoder {
  id
  active
  ensName
  status
  lastRewardRound
  rewardCut
  feeShare
  pricePerSegment
  pendingRewardCut
  pendingFeeShare
  pendingPricePerSegment
  totalStake
  rewards {
    rewardTokens
    round {
      id
      __typename
    }
    __typename
  }
  __typename
}
```

and manually transformed it into this (somehwat monstrous) SQL query:
```sql
with erounds as
 (select e.id, e.data,
        jsonb_build_object('__typename', 'Round', 'id', e.id) as obj
   from entities e
  where e.subgraph='QmfPTP8Kao6kTGRRRDyHKnsQWgDU1ytAqhezJuP3Spq86A'
    and e.entity='Round'),
 erewards as
 (select e.id, e.data,
         jsonb_build_object('__typename', 'Reward',
           'id', e.id,
           'rewardTokens', e.data->'rewardTokens'->>'data') as obj
   from entities e
  where e.subgraph='QmfPTP8Kao6kTGRRRDyHKnsQWgDU1ytAqhezJuP3Spq86A'
    and e.entity='Reward'),
 rewards as
 (select erewards.id, erewards.data,
         (erewards.obj || jsonb_build_object('round', erounds.obj)) as obj
  from erewards left outer join erounds
                on erounds.id = erewards.data->'round'->>'data'),
 etrans as
 (select e.id, e.data,
        jsonb_build_object('__typename', 'Transcoder',
          'active', (e.data->'active'->>'data')::bool,
          'ensName', e.data->'ensName'->>'data',
          'feeShare', e.data->'feeShare'->>'data',
          'id', e.id,
          'lastRewardRound', e.data->'lastRewardRound'->>'data',
          'pendingFeeShare', e.data->'pendingFeeShare'->>'data',
          'pendingPricePerSegment', e.data->'pendingPricePerSegment'->>'data',
          'pendingRewardCut', e.data->'pendingRewardCut'->>'data',
          'pricePerSegment', e.data->'pricePerSegment'->>'data',
          'rewardCut', e.data->'rewardCut'->>'data',
          'status', e.data->'status'->>'data',
          'totalStake', e.data->'totalStake'->>'data') as obj
   from entities e
  where e.subgraph='QmfPTP8Kao6kTGRRRDyHKnsQWgDU1ytAqhezJuP3Spq86A'
    and e.entity='Transcoder')
select etrans.id, (etrans.obj || jsonb_build_object('rewards', json_agg(rewards.obj))) as obj
from etrans left outer join rewards
            on etrans.id = rewards.data->'transcoder'->>'data'
group by etrans.id, etrans.obj;
```

The GraphQL version takes ~ 10s, whereas the above SQL query takes ~ 400ms against the production data. The SQL query also returns data in the form we need to send it back over the wire; the caveat is that I have only manually compared the results of the two, there might be mistakes in my manual translation.

**Comment by Jannis** [comment]
That's great already! I wonder if building the json objects in rust would be faster or slower. My gut feeling is slower.

One thing to keep in mind in the conversion to json is that we maybe represent some value types differently in GraphQL than Postgres does. I'm not sure this is the case though.

**Comment by leodasvacas** [comment]
@lutter Do you plan on keeping the result coercion step in the SQL resolver? It is a step that double-checks that the value returned from the store is consistent with the graphql value (for example checking that a number fits in an i32). It's mostly redundant and if we have sufficient assurance that the store data is consistent it seems unecessary. But the visiting it does of the result tree is very similar to what would be required for custom resolvers, so custom resolution and coercion could be done in the same step in the SQL resolver. Your thoughts?

**Comment by leodasvacas** [comment]
@lutter One graphql detail to remember is that we support `@skip` and `@include`, you can see how they work [here](https://graphql.github.io/graphql-spec/draft/#sec--skip).

**Comment by lutter** [comment]
@leodasvacas yes, my current plan is to return JSON from the query that uses the same types we use in the database to represent these values (e.g., `BigInt` as a string) and leave, at least initially, coercion alone.

**Comment by Jannis** [comment]
This was merged with #1386.
